netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -422,5 +509,115 @@ public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorC
         throws Http2Exception {
       handler.onRstStreamRead(streamId);
     }
+
+    @Override public void onPingAckRead(ChannelHandlerContext ctx, ByteBuf data)
+        throws Http2Exception {
+      Ping p = handler.ping.get();
+      if (p != null) {
+        byte[] bytes = new byte[data.readableBytes()];
+        data.readBytes(bytes);
+        if (p.tryAck(bytes)) {
+          handler.ping.compareAndSet(p, null);
+        }
+        // TODO(jhump): log warning on bad ACK?
+      }
+    }
+  }
+
+  /**
+   * An outstanding ping operation. Accumulates listeners and notifies them when the ping is
+   * acknowledged.
+   */
+  private static class Ping {
```

Comment:
Factored out the two similar `Ping` classes into a single shared on in the `core` module's transport sub-package.